File: leetcode/1-two-sum/solution1.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
function twoSum(nums, target) {
  const map = nums.reduce((res, e, i) => {
    res[e] = i;
    return res;
  }, {});

  for (let i = 0; i < nums.length; i++) {
    const num = nums[i];
    if (
      map.hasOwnProperty(target - num)
      && map[target - num] !== i
    ) {
      return [i, map[target - num]];
    }
  }
}

File: leetcode/1-two-sum/solution1.java
import java.util.HashMap;

class Solution {
  public int[] twoSum(int[] nums, int target) {
    HashMap<Integer, Integer> hm = new HashMap<>();
    for (int i = 0; i < nums.length; i++) {
      hm.put(nums[i], i);
    }

    for (int i = 0; i < nums.length; i++) {
      int remain = target - nums[i];
      if (hm.containsKey(remain) && i != hm.get(remain)) {
        return new int[]{i, hm.get(remain)};
      }
    }
    return new int[2];
  }
}
